### DISCUSSION TITLE
sqlalchemy 1.4 start a unhandle connection while with asyncSession.flush()

### DISCUSSION THREAD

**Discussion by hackrole** [sqlalchemy/sqlalchemy#8949] [Usage Questions]
# the story
I have a fastapi app, it use `sqlalchemy==1.4` for async connection to mysql.
I use factoryboy for prepare mock data, and old sqlalchemy to flush to db.

but I found while the testcase growns, run the testcases start raise `max connection raise`.

# the code demo
```python
# tests/conftest.py
async def db_session():
       engine = create_async_engine("...")
       conn = engine.connect()
       trans = conn.begin()
  
       conn.__class__.force_close = conn.close
       conn.__class__.close = lambda: None
       trans.__class__.force_rollback = trans.rollback
       trans.__class__.rollback = lambda: None
  
       session = create_async_session()
       session.force_close = session.close
       session.close = lambda: No
   
       try:
            yield session
       finally:
            await session.force_close()
            await trans.rollback()
            await conn.close()
            await engine.dispose()
 # tests_something.py
async def test_something(db_session):
          obj = factoryboy.SomeModelFactory.create()
          db_session.add(obj)
          await db_session.flush()  
```

# the reason
after debug, I found the `await db_session.flush()` will create a connection, while will not be closed after the single testcase finish.

so every test will create a unclose connection. while the testcase increase, it finally raise cause  the `mysql max conn` settings.

# what to do?
after check the code, I found no way to close the unhandle connnection.

Is there any advice??

**Comment by CaselIT** [comment]
Hi

> I have a fastapi app, it use sqlalchemy==1.4 for async connection to mysql.

you really should not use 1.4.0. Please update to one of the latest 1.4 releases

> after check the code, I found no way to close the unhandle connnection.
> Is there any advice??

well you are overriding close to be no-op in your example, so maybe don't do that?
